### PR TITLE
User/kiransuren/#86/add gps converter rqtplugin

### PR DESCRIPTION
**Pull request by kiransuren**
Implements a rqt plugin that converts GPS coordinates from DMS to DD and vice versa. Allows for adding and converting multiple coordinates (or removing the last coordinate entered). A separate button allows you to post the coordinate information to GPS topic.

Usage:
-> run: catkin clean, catkin build, source ./devel/setup.bash
-> run: roscore
-> run: rqt --list-plugins (checks if the plugin has been detected, if not don't worry it may still come up with the next step)
->  run: rqt --force-discover
-> select the Plugins tab on the top bar, select GPS Converter from the dropdown menu

Next Steps:
-> Add a gui or at least a display of the current coordinates entered
-> refactor some code (rqt is a bit of a pain to work with and there isn't a lot online about it so there is definitely some places to improve)

### CODE REVIEW

**Comment by kiransuren** [comment]
> Is this a draft? checks not passing

Sorry yes, just changed it to a draft

**Comment by hamzaMahdi** [comment]
Seems the CI cant install the proper versison of qemu for some reason @wmmc88

**Comment by wmmc88** [comment]
> Seems the CI cant install the proper versison of qemu for some reason @wmmc88

looks like github changed one of their hosted images. imma have a fix soon

**Comment by wmmc88** [comment]
@hamzaMahdi @kiransuren fixed issue in #90

**Comment by wmmc88** [comment]
#90 is now merged in. @kiransuren u can rebase off master to get the fix

**Comment by kiransuren** [comment]
@wmmc88 Great thanks!

File: uwrt_rqt_plugins/.vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,61 @@
+{
```
Comment (wmmc88) [inline uwrt_rqt_plugins/.vscode/settings.json:1]: This shouldnt be committed

Comment (kiransuren) [reply uwrt_rqt_plugins/.vscode/settings.json:1]: there's also a .vscode in the meta package, can I just add it to the gitignore?

Comment (wmmc88) [reply uwrt_rqt_plugins/.vscode/settings.json:1]: Yes

Comment (kiransuren) [reply uwrt_rqt_plugins/.vscode/settings.json]: issue fixed

**Review by wmmc88 — COMMENTED** [review]
I'll do a full review later today or tomorrow. 

Could u rename the package to be `uwrt_mars_rover_rqt_gui`? we'll probably add more rqt plugins to this package in the new future. things like buttons to switch and report control modes, finishing integrating mapviz, etc.

Could you also add a short readme about this new package within the new package (akin to whats in uwrt_mars_rover_utils). a lot of people don't know about `rqt --list-plugins` and `rqt --force-discover` so those would be useful to mention.

Add the gui elements to view the list of the current inputted values. (aka do your next steps as part of this pr)

add a rqt perspective file with a rqt perspective that has just this plugin. add a launch file to this package called `rover_gui.launch` that opens an rqt window with the perspective file. ~ALSO, add this launch file to the one in `uwrt_mars_rover_bringup` so that `rover.launch` also launches `rover_gui.launch`.~ **EDIT: ** this actually doesnt make sense to launch in `rover.launch` because it runs on the basestation side.

**Comment by wmmc88** [comment]
I've been doing some testing and here are a couple remarks:
* Coordinates may be provided to us in any of the following:  decimal degrees; degrees decimal minutes;
degrees minutes seconds, BUT we always want to be converting to decimal degrees because thats what ros uses. Our VectorNAV gps will be publishing [NavSatFix](http://docs.ros.org/en/melodic/api/sensor_msgs/html/msg/NavSatFix.html) and I think you should be using that to publish DD coordinates
*  The point of this pr is to provide a useful gui to help with gps input for URC tasks. In URC there are two places we use gps iirc:  the list of 7 gps coordinates for goals in autonomy and a list of x number of gps coordinates provided for extreme terrain traversal. we want to be able to start using the information as soon as possible but also make sure we can edit/change the lists and stuff. so I propose some changes to the scope of this pr:
1. store the coordinates(dd) and their associated names on the parameter server. names examples: could be Leg 1, Leg 2(in autonomy task) or could be toolbox, hammer, rockgarden (in extreme terrain traversal task. maybe the functions to do this should be in utils package?
2. modify your gui so that it takes in any of the 3 gps formats + a name for the coordinate, It should covert to DD and add the name/coordinate pair to the parameter server somehow. probably best to use try to store it as a list of structs or smth. (http://docs.ros.org/en/api/xmlrpcpp/html/classXmlRpc_1_1XmlRpcValue.html). 
3. add functionality to your gui to constantly read the list from the parameters server and display the list of coordinates with their associated names. It should be easy to delete a pair from the list in the gui, add a pair to the list, and reorder the list. 

This should give us some easy ways to later use these coordinates in either our autonomy stack or maybe also add a Go to goal gui element that chooses a point and tells the rover to autonomously navigate towards it. This proposed setup should also allow us to have multiple people have gui nodes running with one ground truth being served (from the ros parameter server that will run on the rover)
 
 maybe we should have a call to discuss this @kiransuren  @azumnanji ?

**Comment by kiransuren** [comment]
For sure, hopping on a call sounds good!

**Comment by wmmc88** [comment]
> For sure, hopping on a call sounds good!

kk I talked to @azumnanji  and he said after the meeting on thursday

**Comment by kiransuren** [comment]
> > For sure, hopping on a call sounds good!
> 
> 
> 
> kk I talked to @azumnanji  and he said after the meeting on thursday

Awesome sounds good

**Pull request closed without merging**